logging/env_logger.h

```diff
@@ -0,0 +1,181 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+//
+// Logger implementation that uses custom Env object for logging.
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include "port/sys_time.h"
+#include <time.h>
+
+#include "monitoring/iostats_context_imp.h"
+#include "rocksdb/env.h"
+#include "rocksdb/slice.h"
+#include "test_util/sync_point.h"
+#include "util/file_reader_writer.h"
+#include "util/mutexlock.h"
+
+namespace rocksdb {
+
+class EnvLogger : public Logger {
+ public:
+  EnvLogger(std::unique_ptr<WritableFile>&& writable_file,
+            const std::string& fname, const EnvOptions& options, Env* env,
+            InfoLogLevel log_level = InfoLogLevel::ERROR_LEVEL)
+      : Logger(log_level),
+        file_(std::move(writable_file), fname, options, env),
+        last_flush_micros_(0),
+        env_(env),
+        flush_pending_(false) {}
+
+  ~EnvLogger() {
+    if (!closed_) {
+      closed_ = true;
+      CloseHelper();
+    }
+  }
+
+  static Status NewEnvLogger(const std::string& fname, Env* env,
+                             std::shared_ptr<Logger>* result) {
```

Comment:
I hope the function is made public under include/rocksdb/. Now users cannot call this function, and they cannot easily call Env::NewLogger() either.

Reply:
Siying, Can you please let me know the file under the include/rocksdb where i should put this method? I will make the change accordingly.

Reply:
For now, maybe env.h?

Reply:
Ok. Moved it back to env.h.  Please review it again.

Reply:
What I see now is that it is back to a function of Env. The problem is that users have to create a Env class to use it, but it takes an `Env` as a parameter. To me it is confusing.

What I suggested is that, we need to keep the function publically available, which means under `include/rocksdb`, this is what users can see after we release the code. Also, it should not be a function of Env. So a workable solution is to create a standalone function in include/rocksdb/env.h.

Reply:
I misunderstood your earlier comment. Made it a standalone function in env.h.
I have implemented it in env.cc as EnvLogger does not have a cc file. If you prefer, i can create a env_logger.cc file to hold the implementation of NewEnvLogger.